src/flowcontrol.rs

```diff
@@ -0,0 +1,214 @@
+    /// of the current window.
+    pub fn should_update_max_data(&self) -> bool {
+        let available_window = self.max_data - self.consumed;
+
+        available_window < (self.window / 2)
+    }
+
+    /// Returns the new max_data limit.
+    pub fn max_data_next(&self) -> u64 {
+        self.consumed + self.window
+    }
+
+    /// Commits the new max_data limit.
+    pub fn update_max_data(&mut self, now: Instant) {
+        self.max_data = self.max_data_next();
+        self.last_update = Some(now);
+    }
+
+    /// Autotune the window size. When there is an another update
+    /// within RTT x 2, bump the window x 1.5, capped by
+    /// max_window.
+    pub fn autotune_window(
+        &mut self, now: Instant, rtt: Duration, max_window: u64,
+    ) {
+        if let Some(last_update) = self.last_update {
+            if now - last_update < rtt * WINDOW_TRIGGER_FACTOR {
+                self.window =
+                    cmp::min(self.window * WINDOW_INCREASE_FACTOR, max_window);
+            }
+        }
+    }
+
+    /// Make sure the lower bound of the window is same to
+    /// the current window.
+    pub fn ensure_window_lower_bound(&mut self, min_window: u64) {
+        self.window = cmp::max(min_window, self.window);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn max_data() {
+        let fc = FlowControl::new(100, 20);
+
+        assert_eq!(fc.max_data(), 100);
+    }
+
+    #[test]
+    fn should_update_max_data() {
+        let mut fc = FlowControl::new(100, 20);
+
+        fc.add_consumed(85);
+        assert_eq!(fc.should_update_max_data(), false);
+
+        fc.add_consumed(10);
+        assert_eq!(fc.should_update_max_data(), true);
+    }
+
+    #[test]
+    fn max_data_next() {
+        let mut fc = FlowControl::new(100, 20);
+
+        let consumed = 95;
+
+        fc.add_consumed(consumed);
+        assert_eq!(fc.should_update_max_data(), true);
+        assert_eq!(fc.max_data_next(), consumed + 20);
+    }
+
+    #[test]
+    fn update_max_data() {
+        let mut fc = FlowControl::new(100, 20);
+
+        let consumed = 95;
+
+        fc.add_consumed(consumed);
+        assert_eq!(fc.should_update_max_data(), true);
+
+        let max_data_next = fc.max_data_next();
+        assert_eq!(fc.max_data_next(), consumed + 20);
+
+        fc.update_max_data(Instant::now());
+        assert_eq!(fc.max_data(), max_data_next);
+    }
+
+    #[test]
+    fn autotune_window() {
+        let w = 20;
+        let mut fc = FlowControl::new(100, w);
+
+        let consumed = 95;
+
+        fc.add_consumed(consumed);
+        assert_eq!(fc.should_update_max_data(), true);
+
+        let max_data_next = fc.max_data_next();
+        assert_eq!(max_data_next, consumed + w);
+
+        fc.update_max_data(Instant::now());
+        assert_eq!(fc.max_data(), max_data_next);
+
+        // Window size should be doubled.
+        fc.autotune_window(Instant::now(), Duration::from_millis(100), 100);
```

Comment:
Can we use fixed instants rather than `now()` just in case some timing issues pop up in CI?

Reply:
thanks. fixed